### PR TITLE
Make "cargo uninstall" uninstall the cwd bins

### LINKED ISSUE #5916: cargo uninstall doesn't uninstall
When in a source directory, I expect `cargo uninstall` to remove the compiled package from `~/.cargo/bin`. It doesn't seem to do so.

In this example I start with no uutils, `cargo install`, have uutils, `cargo uninstall`, still have uutils:

```
brian@DESKTOP-UCV672I:~/coreutils⟫ ls ~/.cargo/bin
basic-http-server  cargo-benchcmp  cargo-fuzz      cargo-upgrade  mdbook  rustc    rustfmt-format-diff  rustup  wasm2es6js                wasm-pack
cargo              cargo-clippy    cargo-generate  comrak         rg      rustdoc  rust-gdb             tokei   wasm-bindgen
cargo-add          cargo-fmt       cargo-rm        git-fmt        rls     rustfmt  rust-lldb            twiggy  wasm-bindgen-test-runner
brian@DESKTOP-UCV672I:~/coreutils⟫ cargo +nightly install
warning: To build the current package use `cargo build`, to install the current package run `cargo install --path .`
  Installing uutils v0.0.1 (file:///home/brian/coreutils)
    Finished release [optimized] target(s) in 0.95s
  Installing /home/brian/.cargo/bin/uutils
brian@DESKTOP-UCV672I:~/coreutils⟫ ls ~/.cargo/bin
basic-http-server  cargo-benchcmp  cargo-fuzz      cargo-upgrade  mdbook  rustc    rustfmt-format-diff  rustup  uutils        wasm-bindgen-test-runner
cargo              cargo-clippy    cargo-generate  comrak         rg      rustdoc  rust-gdb             tokei   wasm2es6js    wasm-pack
cargo-add          cargo-fmt       cargo-rm        git-fmt        rls     rustfmt  rust-lldb            twiggy  wasm-bindgen
brian@DESKTOP-UCV672I:~/coreutils⟫ cargo +nightly uninstall
brian@DESKTOP-UCV672I:~/coreutils⟫ ls ~/.cargo/bin
basic-http-server  cargo-benchcmp  cargo-fuzz      cargo-upgrade  mdbook  rustc    rustfmt-format-diff  rustup  uutils        wasm-bindgen-test-runner
cargo              cargo-clippy    cargo-generate  comrak         rg      rustdoc  rust-gdb             tokei   wasm2es6js    wasm-pack
cargo-add          cargo-fmt       cargo-rm        git-fmt        rls     rustfmt  rust-lldb            twiggy  wasm-bindgen
```

Tested on stable and nightly.

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #5916

Tested with a local build of cargo, using coreutils:

    17:33:57 $ dcargo uninstall
    Removing /Users/dnw/.cargo/bin/uutils

### CODE REVIEW

File: src/cargo/ops/cargo_install.rs [hunk 1]
```diff
@@ -768,13 +781,36 @@ pub fn uninstall_one(
...
+    config: &Config,
+) -> CargoResult<()> {
     let mut to_remove = Vec::new();
     {
-        let result = PackageIdSpec::query_str(spec, metadata.v1.keys())?.clone();
-        let mut installed = match metadata.v1.entry(result.clone()) {
+        let mut installed = match metadata.v1.entry(pkgid.clone()) {
             Entry::Occupied(e) => e,
-            Entry::Vacant(..) => panic!("entry not found: {}", result),
+            Entry::Vacant(..) => panic!("entry not found: {}", pkgid),
```
Comment (ehuss) [inline src/cargo/ops/cargo_install.rs]: This panic happens if you attempt to `cargo uninstall` something that is not installed.

File: src/cargo/ops/cargo_install.rs [hunk 2]
```diff
@@ -768,13 +781,36 @@ pub fn uninstall_one(
...
+
+fn uninstall_cwd(
+    root: &Filesystem,
+    bins: &[String],
+    config: &Config,
+) -> CargoResult<()> {
+    let crate_metadata = metadata(config, root)?;
+    let metadata = read_crate_list(&crate_metadata)?;
+    let source_id = SourceId::for_path(config.cwd())?;
+    let (pkg, _source) = select_path_pkg(None, &source_id, None, config, true)?;
```
Comment (ehuss) [inline src/cargo/ops/cargo_install.rs]: This includes an error message that mentions `--path` and `--git` which don't exist in `uninstall`.

File: src/cargo/ops/cargo_install.rs (hunk 1 above)
Comment (dwijnand) [reply src/cargo/ops/cargo_install.rs]: This is the status quo despite this change. What change are you requesting?

**Comment by dwijnand** [comment]
Thanks for the review, @ehuss! I've address one of your comments, but I'm unsure about the other. 😕

**Comment by dwijnand** [comment]
@ehuss I spent some time reproducing the test case you predicted so now we can iterate on how to handle it better.

**Comment by dwijnand** [comment]
All done from me, ready for re-review.

**Comment by dwijnand** [comment]
This is failing on Travis CI for:

```
---- path::deep_dependencies_trigger_rebuild stdout ----
running `/Users/travis/build/rust-lang/cargo/target/debug/cargo build`
running `/Users/travis/build/rust-lang/cargo/target/debug/cargo build`
running `/Users/travis/build/rust-lang/cargo/target/debug/cargo build`
running `/Users/travis/build/rust-lang/cargo/target/debug/cargo build`
thread 'path::deep_dependencies_trigger_rebuild' panicked at '
Expected: execs
    but: differences:
  0 - |[COMPILING] bar v0.5.0 (file:///Users/travis/build/rust-lang/cargo/target/cit/t871/foo/bar)|
    + |   Compiling baz v0.5.0 (file:///Users/travis/build/rust-lang/cargo/target/cit/t871/foo/baz)|
  1 - |[COMPILING] foo v0.5.0 (file:///Users/travis/build/rust-lang/cargo/target/cit/t871/foo)|
    + |   Compiling bar v0.5.0 (file:///Users/travis/build/rust-lang/cargo/target/cit/t871/foo/bar)|
  2 - |[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]|
    + |   Compiling foo v0.5.0 (file:///Users/travis/build/rust-lang/cargo/target/cit/t871/foo)|
  3 -
    + |    Finished dev [unoptimized + debuginfo] target(s) in 1.71s|
other output:
``', tests/testsuite/support/hamcrest.rs:13:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failures:
    path::deep_dependencies_trigger_rebuild
```

which I'd say is unrelated to this PR. 😕

**Comment by ehuss** [comment]
LGTM, r+ unless anyone else has input.

**Pull request merged**